### PR TITLE
update smokehouse to do deep result comparisons

### PR DESCRIPTION
**Pull request by brendankenny**
Allows smokehouse expectations to be objects now. Smokehouse will walk the expected value until getting to any primitive leaves, comparing to the actual result at every step, failing if there is any difference. This allows for the expected value to not be complete, so it doesn't have to include noisy details (e.g. audit meta info) better checked by unit tests.

I augmented the PWA and offline smoke test expectations files, but it turns out all those older audits don't include anything that interesting in their extended info (at least compared to newer audits). `manifest-theme-color` doesn't even output what the `theme_color` actually is. We can kick off the DBW deep expectations after this PR.

Originally I planned to allow the old expectations format (if it's just a primitive value, assume that it's the `score` property on the actual audit result), but it's really not much work to just explicitly write out the object literal with a `score` property. The literals add considerably to the length of the file, but it means you don't have to keep that special case in mind while writing expectations files. In practice we tend to write these values once and barely touch them again, so it seems fine to be explicit. If others feel strongly about the shorter syntax when you only care about the audit's score, I can add that back in.

This doesn't quite finish #964, as this doesn't include comparator support (e.g. if you want to test the actual value with a regex or a floating point comparison (TTI is ≤ 5000 or whatever)), but that's an easy step after these changes.

### CODE REVIEW

**Comment by brendankenny** [comment]
as an example, if you wanted to test the result of a `manifest-background-color` audit, the full audit result might be

```js
"manifest-background-color": {
  "score": true,
  "displayValue": "",
  "rawValue": true,
  "extendedInfo": {
    "value": "#bababa",
    "formatter": "null"
  },
  "name": "manifest-background-color",
  "category": "Manifest",
  "description": "Manifest contains `background_color`"
},
```

but much of this is boilerplate and better testable at the audit unit test level (we also don't want to update a bunch of test expectations every time we switch `helpText` strings).

So for your expected result you might write
```js
'manifest-background-color': {
  score: true,
  extendedInfo: {
    value: '#bababa'
  }
}
```
This tests *only* the `score` and `extendedInfo.value` values of the audit result, ignoring everything else.

The value is really testing the manifest parser (which is tested elsewhere), so a better use of this is in the [DBW mega test](https://github.com/GoogleChrome/lighthouse/tree/8d0d67144508a52c40467a93a8244f105c97af17/lighthouse-cli/test/fixtures/dobetterweb), where we can check things like violation counts, to make sure all corner cases are correctly found and handled.

File: lighthouse-cli/test/smokehouse/smokehouse.js [hunk 1]
```diff
@@ -133,15 +187,25 @@ function collateResults(actual, expected) {
...
+ * @param {{category: string, equal: boolean, diff: ?Object, actual: boolean, expected: boolean}} assertion
  */
 function reportAssertion(assertion) {
   if (assertion.equal) {
     console.log(`  ${log.greenify(log.tick)} ${assertion.category}: ` +
         log.greenify(assertion.actual));
   } else {
-    console.log(`  ${log.redify(log.cross)} ${assertion.category}: ` +
-        log.redify(`found ${assertion.actual}, expected ${assertion.expected}`));
+    if (!assertion.diff) {
```
Comment (patrickhulce) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: can you flip this if/else? had to pause for what this was meaning `!equal && !diff === 'diff was just missing'` :)

File: lighthouse-cli/test/smokehouse/smokehouse.js [hunk 2]
```diff
@@ -97,9 +97,61 @@ function runLighthouse(url, configPath) {
...
+    return {
+      path,
+      actual,
+      expected
+    };
+  }
+
+  // We only care that all expected's own properties are on actual (and not the other way around).
+  for (const key of Object.keys(expected)) {
+    // Bracket numbers, but property names requiring quotes will still be unquoted.
```
Comment (patrickhulce) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: why? do we use this for anything other than informing the user?

Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: yeah, this is just for nice formatting. `myArray.1.value` looked weird :)

**Comment by brendankenny** [comment]
> What are your thoughts on allowing the expectations to be functions and passing an assert or some similar chai expect object (with maybe a non-function type defaulting to .score equality? Just don't want to find ourselves basically writing a new test comparison assertion library for a lighthouse-specific expectations when there are some pretty good ones out there.

The main issue I had was I couldn't find anyone doing one-sided deep equality checking, so that if you don't include something on `expected` it didn't care if it was on `actual`. That leaves you writing a comparison that checks the subset of values you're interested in yourself, checking `actual.score`, then `actual.extendedInfo`, then `actual.extendedInfo.value`. To me the declarative form of that, which directly mirrors the actual output LH gives for a run, was simpler and more readable.

This isn't so bad (if this is what you mean):

```js
// results defined up here somewhere
it('correctly does whatever with manifest-background-color', () => {
  const manifestBgColorResult = results.audits['manifest-background-color'];
  assert.strictEqual(manifestBgColorResult.score, true);
  assert(manifestBgColorResult.extendedInfo);
  assert.strictEqual(manifestBgColorResult.extendedInfo.value, '#bababa');
});
```

but

```js
'manifest-background-color': {
  score: true,
  extendedInfo: {
    value: '#bababa'
  }
}
```

scans a lot easier to me.

> It's not super clear to me how this extends to arbitrary comparisons we might want to make (existence of something in a list for example)

basically if when walking the expected result the expected value is a function, we call the function [here](https://github.com/GoogleChrome/lighthouse/pull/1450/files#diff-736bf5119b0a1117d164a95d24ef4234R114) with the actual result, and it can do what it likes. That could return a boolean or, if we wanted, throw an assertion error (which would make deciding what to show in expected vs found a lot easier). This could also be extended to support regexp objects (test against actual) and the numeric string conditionals some people like (`">=5000"` or whatever, [now used by pwmetrics](https://github.com/paulirish/pwmetrics#defining-expectations))

File: lighthouse-cli/test/smokehouse/smokehouse.js (hunk 1 above)
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: done

File: lighthouse-cli/test/smokehouse/smokehouse.js (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: haha gotcha

**Comment by patrickhulce** [comment]
> The main issue I had was I couldn't find anyone doing one-sided deep equality checking
Yeah none that I know of either, but `assert` is pretty limited here and `chai`'s nested deep property checking doesn't make it that bad IMO, but given

> basically if when walking the expected result the expected value is a function, we call the function here with the actual result, and it can do what it likes. 

Basically what I was looking for, SGTM!

**Comment by ebidel** [comment]
I'll give this a look on the way home.

File: lighthouse-cli/test/smokehouse/pwa-expectations.js [hunk 3]
```diff
@@ -25,104 +25,266 @@ module.exports = [
...
+        score: true
+      },
+      'manifest-background-color': {
+        score: true,
+        extendedInfo: {
+          value: '#2196F3'
+        }
+      },
+      'manifest-theme-color': {
+        score: true
```
Comment (ebidel) [inline lighthouse-cli/test/smokehouse/pwa-expectations.js]: does this one have a color val we can check?

File: lighthouse-cli/test/smokehouse/pwa-expectations.js [hunk 4]
```diff
@@ -25,104 +25,266 @@ module.exports = [
...
+      },
+      'manifest-icons-min-192': {
+        score: true
+      },
+      'manifest-icons-min-144': {
+        score: true
+      },
+      'manifest-name': {
```
Comment (ebidel) [inline lighthouse-cli/test/smokehouse/pwa-expectations.js]: I assume the name and urls don't have a val we can check?

File: lighthouse-cli/test/smokehouse/smokehouse.js (hunk 2 above)
Comment (ebidel) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: lol `myArray.1.value` is what polymer uses for it's API.

File: lighthouse-cli/test/smokehouse/smokehouse.js [hunk 5]
```diff
@@ -97,9 +97,61 @@ function runLighthouse(url, configPath) {
...
+ * @param {*} expected
+ * @return {({path: string, actual: *, expected: *}|null)}
+ */
+function findDifference(path, actual, expected) {
+  // Strict equality check, plus NaN equivalence.
+  if (Object.is(actual, expected)) {
+    return null;
+  }
+
+  // if they aren't both an object, reject.
```
Comment (ebidel) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: not sure what "reject" means.

Comment (ebidel) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: Maybe something other than "reject". ...Can't create diff if actual and expected are both not objects.

File: lighthouse-cli/test/smokehouse/pwa-expectations.js [hunk 6]
```diff
@@ -25,104 +25,266 @@ module.exports = [
...
-      'manifest-background-color': true,
-      'manifest-theme-color': true,
-      'manifest-icons-min-192': true,
-      'manifest-icons-min-144': true,
-      'manifest-name': true,
-      'manifest-short-name': true,
-      'manifest-start-url': true,
-      // 'cache-start-url': true
+      'is-on-https': {
+        score: true,
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/pwa-expectations.js]: done

File: lighthouse-cli/test/smokehouse/pwa-expectations.js (hunk 3 above)
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/pwa-expectations.js]: no, since it was written before `extendedInfo` it doesn't include it

File: lighthouse-cli/test/smokehouse/pwa-expectations.js (hunk 4 above)
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/pwa-expectations.js]: > I assume the name and urls don't have a val we can check?

nope, none of these older ones do

File: lighthouse-cli/test/smokehouse/smokehouse.js (hunk 5 above)
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/smokehouse.js]: Good point. Is this more understandable?

**Comment by brendankenny** [comment]
updated

**Comment by ebidel** [comment]
⬇️ 🕳 🚬 🔢 💯
